### ISSUE TITLE
[lsp] Don't run the generator for open files

### ISSUE THREAD

**Issue by spoenemann** [eclipse/xtext-core#373]
Currently the generator is triggered for every build, even when a file is opened. The generator should be suppressed for builds triggered by `didOpen`, any also for files that are open and dirty (a `didChange` has been received).

**Comment by svenefftinge** [comment]
I reconsidered this and now think we should disable the generator always, because we would need two distinct builds to keep file system state separated from in-memory state.
As the LSP is about editing, we should not put both things into one process by default, but rather recommend to run a regular language server without code generation, and have another process that watches the file system for doing code generation.

So for LSP we should bind a NEVER impl of `IShouldGenerate`.

**Comment by cdietrich** [comment]
i prefer at least on option to enable it

**Comment by svenefftinge** [comment]
That would mean we trigger file changes because of in-memory document changes, which is not a good idea.

**Comment by cdietrich** [comment]
maybe moving to org.eclipse.xtext.ide.server.LanguageServerImpl.didSave(DidSaveTextDocumentParams) ?

**Comment by svenefftinge** [comment]
No, the problem is, that we only have one representation per URI and if it is opened, we will use the in-memory state, which would always be wrong for code generation. 
The only way to solve this is to have two representations (file and in-memory) which boils down to two builds and two resourcesets. In that case I don't see an advantage of doing it as part of the language server.

**Comment by cdietrich** [comment]
hmmm but if we save the represenations are equals?
otherwise you would always have to introduce a custom command to generate?

**Comment by svenefftinge** [comment]
I can save one file but not the other, which still might be affected.

**Comment by svenefftinge** [comment]
We could say we generate only if there are no open files with changes, but that is counter intuitive for the user.

**Comment by svenefftinge** [comment]
re: introduce a command 
I would not do it with a command, but by running it the command line, npm or gradle.

**Comment by cdietrich** [comment]
hmmm am not convinced if we want to replace e.g. eclipse integration with lsp4e one day for example

**Comment by cdietrich** [comment]
it actually uses org.eclipse.core.runtime.IProgressMonitor / NullProgressMonitor which is in equinox common but that wuld not help you neither correct?

**Comment by cdietrich** [comment]
am not sure if the current IncrementalBuilder is too "stupid" to handle such a case.
if we could tell in build request if a generation should happen.
then we can do the generation in initial build and on save (which we currently ignore in lsp)
and do not generator during document changes
(am not sure what about side usecases like file gets green during change on other file)

**Comment by svenefftinge** [comment]
As said before, we could only generate if all files are in sync with the file system, because a just saved file could reference a dirty document that affects the code generation.
I imagine that it will be surprising or at least non-intuitive for users if we only generate once all editors are saved. 
Depending on the actual language and code generator this might not be an issue. I just would like to have a sensible default behavior.
